File: test-suite/src/test/java/io/micronaut/mqtt/docs/serdes/ProductInfoSerDesSpec.java
package io.micronaut.mqtt.docs.serdes;

import io.micronaut.context.ApplicationContext;
import io.micronaut.mqtt.test.AbstractMQTTTest;
import org.junit.jupiter.api.Test;

import static java.util.concurrent.TimeUnit.SECONDS;
import static org.awaitility.Awaitility.await;

public class ProductInfoSerDesSpec implements AbstractMQTTTest {

    @Test
    void testUsingACustomSerDes() {
        ApplicationContext applicationContext = startContext();

// tag::producer[]
        ProductClient productClient = applicationContext.getBean(ProductClient.class);
        productClient.send(new ProductInfo("small", 10L, true));
        productClient.send(new ProductInfo("medium", 20L, true));
        productClient.send(new ProductInfo(null, 30L, false));
// end::producer[]

        ProductListener productListener = applicationContext.getBean(ProductListener.class);

        try {
            await().atMost(5, SECONDS).until(() ->
                    productListener.messages.size() == 3 &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 10L) &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 20L) &&
                    productListener.messages.stream().anyMatch(pi -> pi.getCount() == 30L)
            );
        } finally {
            applicationContext.close();
        }
    }
}

File: test-suite-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/serdes/ProductInfoSerDesSpec.kt
package io.micronaut.mqtt.docs.serdes

import io.micronaut.mqtt.AbstractMQTTTest
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class ProductInfoSerDesSpec : AbstractMQTTTest() {

    @Test
    fun testProductInfoSerDes() {
        val applicationContext = startContext()

// tag::producer[]
        val productClient = applicationContext.getBean(ProductClient::class.java)
        productClient.send(ProductInfo("small", 10L, true))
        productClient.send(ProductInfo("medium", 20L, true))
        productClient.send(ProductInfo(null, 30L, false))
// end::producer[]

        val listener = applicationContext.getBean(ProductListener::class.java)

        try {
            await().atMost(5, TimeUnit.SECONDS).until {
                listener.messages.size == 3 &&
                        listener.messages.any { p -> p.size == "small" && p.count == 10L && p.sealed } &&
                        listener.messages.any { p -> p.size == "medium" && p.count == 20L && p.sealed } &&
                        listener.messages.any { p -> p.size == null && p.count == 30L && !p.sealed }
            }
        } finally {
            applicationContext.close()
        }
    }
}
